references/optical_flow/train.py

```diff
@@ -0,0 +1,331 @@
+
+    if args.resume is not None:
+        d = torch.load(args.resume, map_location="cpu")
+        if args.map_orig_to_ours:
+            d = utils.map_orig_to_ours(d)
+        model.load_state_dict(d, strict=True)
+
+    if args.train_dataset is None:
+        # Set deterministic CUDNN algorithms, since they can affect epe a fair bit.
+        torch.backends.cudnn.benchmark = False
+        torch.backends.cudnn.deterministic = True
+        validate(model, args)
+        return
+
+    print(f"Parameter Count: {sum(p.numel() for p in model.parameters() if p.requires_grad)}")
+
+    torch.backends.cudnn.benchmark = True
+
+    model.train()
+    if args.freeze_batch_norm:
+        utils.freeze_batch_norm(model.module)
+
+    train_dataset = get_train_dataset(args.train_dataset, args.dataset_root)
+
+    sampler = torch.utils.data.distributed.DistributedSampler(train_dataset, shuffle=True, drop_last=True)
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        sampler=sampler,
+        batch_size=args.batch_size,
+        pin_memory=True,
+        num_workers=args.num_workers,
+    )
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay, eps=args.adamw_eps)
+
+    scheduler = torch.optim.lr_scheduler.OneCycleLR(
+        optimizer=optimizer,
+        max_lr=args.lr,
+        total_steps=args.num_steps + 100,
+        pct_start=0.05,
+        cycle_momentum=False,
+        anneal_strategy="linear",
+    )
+
+    logger = utils.MetricLogger()
+
+    done = False
+    current_epoch = current_step = 0
+    while not done:
+        sampler.set_epoch(current_epoch)  # needed, otherwise the data loading order would be the same for all epochs
+        print(f"EPOCH {current_epoch}")
+
+        for data_blob in logger.log_every(train_loader):
+
+            optimizer.zero_grad()
+
+            image1, image2, flow, valid_flow_mask = (x.cuda() for x in data_blob)
+            flow_predictions = model(image1, image2, num_flow_updates=args.num_flow_updates)
+
+            loss, metrics = utils.sequence_loss(flow_predictions, flow, valid_flow_mask, args.gamma)
+            loss.backward()
+
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1)
+
+            optimizer.step()
+            scheduler.step()
+
+            logger.update(**metrics)
+
+            current_step += 1
+
+            if current_step == args.num_steps:
+                done = True
+                break
+
+        # Note: we don't sync the SmoothedValues across processes, so the printed metrics are just those of rank 0
+        print(f"Epoch {current_epoch} done. ", logger)
+
+        current_epoch += 1
+
+        if args.rank == 0:
+            torch.save(model.state_dict(), Path(args.output_dir) / f"{args.name}_{current_epoch}.pth")
+            torch.save(model.state_dict(), Path(args.output_dir) / f"{args.name}.pth")
```

Comment:
We should definitely do this. I think it's worth refactoring the script to have same functionality and structure as other reference scripts. Moreover we will need to link the ref scripts with the model prototype and add the `--weights` feature switch.

@NicolasHug do you mind creating an issue for all the above so that we dont forget?

Reply:
Opened https://github.com/pytorch/vision/issues/5056